### DISCUSSION TITLE
How to use the code generator with `quarkus-rest-client-reactive-jackson` scope set to `provided`.

### DISCUSSION THREAD

**Discussion by koplas** [quarkiverse/quarkus-openapi-generator#618] [Q&A]
Hi,
Currently, it is not possible to use the code generator with `quarkus-rest-client-reactive-jackson` when the scope of the dependency is set to `provided`, as this fails with:
```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.6.3:generate-code (default) on project iam: Quarkus code generation phase has failed: InvocationTargetException: You need to add io.quarkus:quarkus-rest-client-jackson to your dependencies.
```

The reason is that quarkus extensions export their capabilities, and this generator checks for the capability of the rest-client and fails incorrectly, despite the extension being later present at runtime.

Thanks!

**Comment by hbelmiro** [comment]
Hi @koplas.
Do you mind sharing a reproducer?

**Comment by koplas** [comment]
Here is a reproducer:
```shell script
git clone https://github.com/koplas/quarkus-openapi-repro.git
cd quarkus-openapi-repro
mvn package
```
This repro only contains the generator, a rest client, and the starter code.

**Comment by mcruzdev** [comment]
Hi @hbelmiro and @koplas, after researching here...

When the `io.quarkus:quarkus-rest-client-jackson` is `provided` the only capability found is:

```
Capability: io.quarkus.cdi
Capability: io.quarkus.rest
Capability: io.quarkus.resteasy.reactive
Capability: io.quarkus.vertx
Capability: io.quarkus.vertx.core
Capability: io.quarkus.vertx.http
```

without `provided`:

```shell
Capability: io.quarkus.cdi
Capability: io.quarkus.rest
Capability: io.quarkus.resteasy.reactive
Capability: io.quarkus.vertx
Capability: io.quarkus.vertx.core
Capability: io.quarkus.vertx.http
Capability: io.quarkus.resteasy.json.jackson.client
```

I think that we have two options:

- make more research to see if there is a way to get capabilities for `provided` scope
- look the resolved dependencies, doing something like this:
```java
// use of method
isCompileOnlyResolvedDependencyPresent(context, "io.quarkus:quarkus-rest-client-reactive-jackson")

// find resolved dependencies with `provided` scope.
private static boolean isCompileOnlyResolvedDependencyPresent(CodeGenContext context, String ga) {
    return StreamSupport
            .stream(context.applicationModel().getDependenciesWithAnyFlag(DependencyFlags.COMPILE_ONLY).spliterator(),
                    false)
            .map(resolvedDependency -> resolvedDependency.getGroupId() + ":" + resolvedDependency.getArtifactId())
            .anyMatch(ga::equals);
}
```

**Reply by hbelmiro** [reply]
I don't see any problem in using the second option. But we could ask in Quarkus Zulip if there's a proper way to get capabilities for provided scope.

**Reply by mcruzdev** [reply]
I agree with you, I was creating a topic on `dev` channel

**Reply by mcruzdev** [reply]
Waiting here: 

https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Find.20capabilities.20with.20dependency.20with.20provided.20scope.3F/near/420652536

**Comment by mcruzdev** [comment]
Hi @koplas, what is your use case to use `provided` scope with the extesion?

**Reply by koplas** [reply]
I use the `quarkus-openapi-generator` in a Keycloak extension. Currently, it is not possible to include the `quarkus-rest-client-reactive-jackson` dependency without resulting in conflicts. So, I set this dependency to `provided` and included the dependency in the Keycloak runtime. It is more of a workaround for https://github.com/keycloak/keycloak/issues/25589 and https://github.com/keycloak/keycloak/issues/20926.
Even without this issue, it could help write extensions where the Quarkus runtime already provides the required dependencies. Of course, for standalone applications, this provides no additional benefit.

**Reply by mcruzdev** [reply]
Thank you @koplas for the explaining that!

**Reply by mcruzdev** [reply]
As this extension do things at compile time, IMHO this can be a issue. WYT @hbelmiro @ricardozanini?

**Reply by hbelmiro** [reply]
> IMHO this can be a issue.

What do you mean by "this" @mcruzdev?

**Reply by mcruzdev** [reply]
I think that we can provide it to our user, I will create the issue hre!
